runtime/libcn/src/utils.c

```diff
@@ -96,10 +96,30 @@ void ghost_stack_depth_incr(void) {
     print_error_msg_info();
 }
 
+
+#define FMT_PTR "\x1b[33m%#lx\x1b[0m"
+// #define KMAG  "\x1B[35m"
+#define FMT_PTR_2 "\x1B[35m%#lx\x1B[0m"
+
 void ghost_stack_depth_decr(void) {
     cn_stack_depth--;
     // update_error_message_info(0);
     print_error_msg_info();
+    // TODO - revisit this Dhruv over-confident change
+    hash_table_iterator it = ht_iterator(cn_ownership_global_ghost_state);
+    printf("CN pointers leaked at (%ld) stack-depth: ", cn_stack_depth);
+    _Bool leaked = false;
+    while (ht_next(&it)) {
+        uintptr_t *key = it.key;
+        int *depth = it.value;
+        _Bool fine = *depth <= cn_stack_depth;
+        if (!fine) {
+            leaked = true;
+            printf(FMT_PTR_2 " (%d),", *key, *depth);
+        }
+    }
+    printf("\n");
+    c_ghost_assert(convert_to_cn_bool(!leaked));
```

Comment:
The unfortunate issue with the current assertion setup is that there is a `cn_assert` function for asserting whether facts about the _CN runtime spec_ are true, and a distinct `c_ghost_assert` function for asserting whether facts about _C ownership_ are true or not. The printed error messages differ depending on which one is called. This falls apart when talking about the global ghost state as a whole, since the hash map `ownership_ghost_state` is currently being used for both types of ownership checking. However, I think it's alright to keep this in for now, and this should probably be a future discussion point regarding the design of the global ghost state and what it should store about the addresses it has mapped.